Upgrade systemlink (#45)
ni/systemlink-operations-handbook: handbook/upgrade/upgrade.md

# Upgrading and Migrating SystemLink Server

Upgrading to a new version or migrating between SystemLink Servers is a common practice in the lifecycle of any SystemLink deployment. The goal of this chapter is to complete this operation while retaining a fully operable SystemLink Server. This is critical to mitigate against issues encountered during an upgrade or migration that prevent the new instance of SystemLink Server from being used.

!!!important "Using Virtual Machines for SystemLink Deployments"

>     It is recommended to run SystemLink Server, attached file stores, and databases within virtual machines. Virtual machines simplify the creation and restoration of backups. The use of other tools to facilitate the backup of bare metal machines, e.g.non-virtual machines, is outside the scope of this chapter.

Reviewer: ```suggestion
    It is recommended to run SystemLink Server, attached file stores, and databases within virtual machines. Virtual machines simplify the creation and restoration of backups. The use of other tools to facilitate the backup of bare metal machines, e.g. non-virtual machines, is outside the scope of this chapter.
```

>     It is recommended to run SystemLink Server, attached file stores, and databases within virtual machines. Virtual machines simplify the creation and restoration of backups. The use of other tools to facilitate the backup of bare metal machines, e.g.non-virtual machines, is outside the scope of this chapter.

Reviewer 2: For this section, I would simplify down to the two main points: This chapter is about upgrading and migrating while keeping SystemLink operational and NI recommends using VMs. 

Here is a super minimalist rewrite possibility:

This chapter covers how to maintain a fully operable SystemLink Server while upgrading to a new version of SystemLink or migrating between SystemLink servers.

To simplify the creating and restoring backups, NI recommends running SystemLink Server, attached file stores, and databases within virtual machines. Backing up physical machines is outside the scope of this chapter.

>     It is recommended to run SystemLink Server, attached file stores, and databases within virtual machines. Virtual machines simplify the creation and restoration of backups. The use of other tools to facilitate the backup of bare metal machines, e.g.non-virtual machines, is outside the scope of this chapter.

Author: I've incorporated the minimalist recommendation

## Assumptions and Prerequisites

> - A server running SystemLink 2021 R1 or greater

Reviewer 3: Does this need to be strengthened? nislmigrate will not work with older versions at this time.

Author: I added another sentence to the bullet stating this point.

- Familiarity with installing and setting up a SystemLink Server

    - [Setting up a SystemLink Server](https://www.ni.com/documentation/en/systemlink/latest/setup/)

    - [Configuring NI Web Server](https://www.ni.com/documentation/en/ni-web-server/latest/manual/configuring-ni-web-server/)

        - [Sign on with LDAP](/ldap/ldap)

        - [Single Sign-on with OpenID Connect](/openid-connect/openid-connect/)

    - Data Store Configuration

        - [MongoDB](/data-stores/mongodb/)

        - [Connecting to a Remote Mongo Database](https://www.ni.com/documentation/en/systemlink/latest/setup/remote-mongo-database/)

        - [Connecting to a Remote File Share](https://www.ni.com/documentation/en/systemlink/latest/setup/remote-file-share/)

        - [Uploading Files to Amazon Simple Storage Service (S3)](https://www.ni.com/documentation/en/systemlink/latest/data/uploading-files-to-amazon-s3/)

        - [Connecting to a Remote PostgreSQL Database](https://www.ni.com/documentation/en/systemlink/latest/setup/remote-postgres-databse/)

- Recommended

>     - All servers, file stores, and databases running in virtual machines

Reviewer 2: Virtual machines to run all servers, file stores, a databases

    - Removable storage volume or device

## Preparing for Upgrade or Migration

> Prior to attempting any upgrade or migration you should backup your server and any remote data stores. Failure to create backups increases the risk of unplanned downtime if issues occur during the upgrade or migration process.

Reviewer 2: Before you begin, backup your server and any remote data to avoid unplanned downtime if issues occur during the upgrade or migration process.

!!!note "Single Node versus Multi Node Configurations"
    **Single node** refers to the SystemLink Server configuration where the application server, file storage, and databases used by SystemLink are all installed on the same Windows server. This is the default configuration for SystemLink Server.

>     **Multi node** refers to configurations where the file store or databases used by SystemLink are running on distinct hardware from the SystemLink application server. This is the recommended configuration for all production deployments.

Reviewer 2: I agree with JD's suggestion and would add that "hardware distinct from" flows a little better than "distinct hardware from."

### NI-SystemLink-Migration Command Line Utility

> The workflows for upgrades and migrations make use of the SystemLink command line migration tool, `nislmigrate`. Refer to the readme at the [NI-SystemLink-Migration](https://github.com/ni/NI-SystemLink-Migration) GitHub repo for details on installing and using this tool.

Reviewer 4: I like the idea of linking to https://pypi.org/project/nislmigrate/, since it has a link to the github repo for those who want to dig more, but is more user facing than the repo.

> The workflows for upgrades and migrations make use of the SystemLink command line migration tool, `nislmigrate`. Refer to the readme at the [NI-SystemLink-Migration](https://github.com/ni/NI-SystemLink-Migration) GitHub repo for details on installing and using this tool.

Reviewer 2: Should be "make" since the subject is the plural "workflows."

> The workflows for upgrades and migrations make use of the SystemLink command line migration tool, `nislmigrate`. Refer to the readme at the [NI-SystemLink-Migration](https://github.com/ni/NI-SystemLink-Migration) GitHub repo for details on installing and using this tool.

Author: done!

!!!note "Argument Flags for `nislmigrate`"
    `nislmigrate` supports capturing data from individual services as well as all installed services using the `--all` argument flag. For brevity `--all` is used most workflows in this chapter. Depending on your needs you may replace the `--all` argument flag with one or more of the individual service argument flags.

> ## Recommended Upgrade and Migrations Workflows for your deployment

Reviewer: ```suggestion
## Recommended Upgrade and Migration Workflows for your deployment
```

> This chapter recommends workflows for the following upgrade and migration scenarios. You should plan for some downtime of your SystemLink Server. By following these recommendations this downtime can be minimized.

Reviewer 2: While you should plan for some downtime of your SystemLink Server, you can minimize that downtime by following these recommendations.

- [Single Node Upgrade](#single-node-upgrade)
- [Single Node to Multi Node Migration](#single-node-to-multi-node-migration)
    - [Single Node to Multi Node with MongoDB](#single-node-to-multi-node-with-mongodb)
    - [Single Node to Multi Node with File Storage](#single-node-to-multi-node-with-file-storage)
    - [Single Node to Multi Node with PostgreSQL](#single-node-to-multi-node-with-postgresql)
- [Upgrading Multi node configurations (MongoDB, PostgreSQL, File Share/S3)](#upgrading-multi-node-configurations)
- [Seamless cut-over](#seamless-cut-over)

!!!note "Storing Captured Data"

>     For upgrades and migrations for a SystemLink Server with a small amount of data you may be able to use `nislmigrate` to backup your data to the local file system of your SystemLink Server. For SystemLink Servers with a large amount of data be aware `nislmigrate` could exhaust the local storage on the server. This can mitigated by using an attached volume. **This document assumes this will be needed and refers to this attached volume as `D:\` in each workflow.**

Reviewer 2: Can we be more specific about what constitutes small/large amounts of data? 

Rewrite for active voice:

If you have a SystemLink Server with a small amount of data, you can use 'nislmigrate' to backup your data to the local file system of your SystemLink Server. For SystemLink Servers with a large amount of data, you may need to use an attached volume since 'nislmigrate' may exhaust the local storage. This document assumes that you are using an attached volume and refers to the volume as 'D:\' in each workflow.

Author: I included your recommended wording. Thank you. 

Per small vs large amount of data. This is dependent on how much data they currently have and how much free disk space they have on their server. This can be a bit squishy depending on the scenario and therefore I would rather not define specifically (in GBs or otherwise) what is a small versus large amount of data.

## Single Node Upgrade

> This workflow describes steps to upgrade a single node deployment of SystemLink Server.

Reviewer 2: This workflow explains how to upgrade a single node deployment of SystemLink Server.

> The SystemLink NI Package Manager Installers (NIPM) supports in-place upgrades of SystemLink Server. An in place upgrade is one where the upgrade is run directly on your current SystemLink Server. For production deployments of SystemLink Server this is not recommended. If you pursue this option it is highly recommended you backup your server prior to the upgrade. This backup is an essential fallback should an issue occur during upgrade that leaves your SystemLink Server in a bad state.

Reviewer 3: ```suggestion
The NI Package Manager (NIPM) installer for SystemLink supports in-place upgrades of SystemLink Server. An in-place upgrade is one where the upgrade is run directly on your current SystemLink Server. NI does not recommend in-place upgrades for production deployments of SystemLink Server. If you pursue this option it is highly recommended that you backup your server prior to the upgrade. This backup is an essential fallback should an issue occur during upgrade that leaves your SystemLink Server in a bad state.
```

Reviewer 2: Though the NI Package Manager installer for SystemLink supports in-place upgrades where the upgrade runs directly on your current SystemLink Server, NI does not recommend this option. If you choose this option, ensure that you backup your server before beginning the upgrade.

> The SystemLink NI Package Manager Installers (NIPM) supports in-place upgrades of SystemLink Server. An in place upgrade is one where the upgrade is run directly on your current SystemLink Server. For production deployments of SystemLink Server this is not recommended. If you pursue this option it is highly recommended you backup your server prior to the upgrade. This backup is an essential fallback should an issue occur during upgrade that leaves your SystemLink Server in a bad state.

Reviewer: ```suggestion
The SystemLink NI Package Manager Installers (NIPM) supports in-place upgrades of SystemLink Server. An in-place upgrade is one where the upgrade is run directly on your current SystemLink Server. For production deployments of SystemLink Server this is not recommended. If you pursue this option, it is highly recommended you backup your server prior to the upgrade. This backup is an essential fallback should an issue occur during upgrade that leaves your SystemLink Server in a bad state.
```

> For single node upgrades, NI recommends pairing the upgrade with a migration. This mitigates risks of issues during the upgrade by ensuring your original SystemLink Server remains in an operable state.

Reviewer 2: For single node upgrades, NI recommends upgrading and migrating at the same time to mitigate issues during the upgrade by ensuring your original SystemLink Server remains operable.

1. Backup your SystemLink Server.

> 1. Attach a volume to store data captured by `nislmigrate`. This will be referenced as `D:\`.

Reviewer 3: The migration will contain secret (private key) and sensitive data (the user's production data). We try to protect the secret data, but I wouldn't consider this bulletproof, and we don't protect the sensitive data. Is there an appropriate warning we should give about proper handling of the migration data in transit and after the upgrade is complete?

1. Install `nislmigrate`.

> 1. Run the command `nislmigrate capture --all --secret <your secret> --dir D:\migration`.

Reviewer 3: Will readers understand what the secret is in this context?

Author: See next comment

    !!!note

>         While `nislmigrate` runs all SystemLink services are stopped. Plan ahead for this downtime of your SystemLink server. When the capture is completed, SystemLink services will be restarted automatically.

Reviewer 3: Should there be a general preparation step for common setup and information like this which is common to all paths?

Author: After writing all the workflows down I was wondering the same thing. I'm going to work through the rest of the feedback and then determine if there is a good way to reduce this repetition.

1. Detach `D:\`.

    !!!note
        At this point in the workflow your original SystemLink Server is in an operable state. Be aware if any new data is created or ingested by SystemLink Server at this time it will not be available to your new server.

> 1. Provision a new Window server for SystemLink.

Reviewer 3: ```suggestion
1. Provision a new Windows server for SystemLink.
```

1. Install and configure the new version of SystemLink Server.

1. Install `nislmigrate` on your new SystemLink Server.

1. Attach the `D:\` volume used to capture data from your original SystemLink Server.

1. Run the command `nislmigrate restore --all --secret <your secret> --dir D:\migration`.

> 1. Verify your new SystemLink Server has all the expected migrated data.

Reviewer 3: What about TDMS? No TDMS data will have been migrated with this procedure.

Author: We'll migrate TDMS files,  but I believe you're referring the rest of the TDM resources. I added an  !!!important in the section introducing the migration tool to refer to PyPi for the list of services it currently supports.

## Single Node to Multi Node Migration

> This section describes workflows used when you intend to upgrade or migrate from a single node SystemLink Server configuration to a multi node SystemLink Server configuration that makes use of dedicated servers for MongoDB, PostgreSQL, or file storage. While not required, these migration workflows include provisioning a new SystemLink Server. This is done to reduce risk and ensure your original SystemLink Server is always in an operable state.

Reviewer 2: I would just mark the step about provisioning as optional rather than calling it out here and having the user jump around mentally before they start the process.

Suggested rewrite:

This section describes workflows to prepare to upgrade or migrate from a single node SystemLink Server configuration to a multi node SystemLink Server configuration that makes use of dedicated servers for MongoDB, PostgreSQL, or file storage.

Author: I like the suggestions. Included your rewrite and put the comment about options in the context of the workflow step.

### Single Node to Multi Node with MongoDB

> This workflow describes steps to upgrade a single node deployment of SystemLink Server to a multi node deployment where the MongoDB instance used by SystemLink is hosted on its own server or replica set.

Reviewer 2: I'm guessing this would be a style choice you might not want to go to the trouble of implementing everywhere but "Complete the following steps to" is more active and concise than "This workflow describes steps to."

Author: Great suggestion. I've incorporated this in each section.

1. Backup your SystemLink Server.

1. Attach a volume to store data captured by `nislmigrate`. This will be referenced as `D:\`.

1. Install `nislmigrate`.

1. Run the command `nislmigrate capture --all --secret <your secret> --dir D:\migration`.

1. Detach `D:\`.

>     !!!note

Reviewer 2: Since your original SystemLink Server is still operable, any new data created or consumed by SystemLink Server at this time will not be available to your new server.

Author: I've went ahead in used this wording on the all the notes that reference this info.

        At this point in the workflow your original SystemLink Server is in an operable state. Be aware if any new data is created or ingested by SystemLink Server at this time it will not be available to your new server.

1. Provision a new Windows server for SystemLink.

> 1. Provision a new MongoDB server or replica set.

Reviewer 3: Should we link to the relevant section of this document on MongoDB configuration?

1. Install and configure the same or a newer version of SystemLink Server.

1. Configure SystemLink to use the newly created MongoDB server or replica set.

1. Install `nislmigrate` on your new SystemLink Server.

1. Attach the `D:\` volume used to capture data from your original SystemLink server.

1. Run the command `nislmigrate restore --all --secret <your secret> --dir D:\migration`.

1. Verify your new SystemLink Server has all the expected migrated data.

### Single Node to Multi Node with File Storage

> This workflow describes steps to upgrade a single node deployment of SystemLink Server to a multi node deployment where the file storage instance used by SystemLink is hosted on a a dedicated NAS, SAN, or AWS S3.

Reviewer: ```suggestion
This workflow describes steps to upgrade a single node deployment of SystemLink Server to a multi node deployment where the file storage instance used by SystemLink is hosted on a dedicated NAS, SAN, or AWS S3.
```

1. Backup your SystemLink Server.

1. Attach a volume to store data captured by `nislmigrate`. This will be referenced as `D:\`.

1. Install `nislmigrate`.

1. Run the command `nislmigrate capture --all --secret <your secret> --dir D:\migration`.

> 1. Copy the files from the original file storage location to the new file store.

Reviewer 2: Copy the files from the original file storage location to the new file store using one of the following options.

Table with migration location and method here.

Author: Done. Let me know if the table covers your guidance as expected.

    !!!note
        This file copy/move is not facilitated by `nislmigrate` directly. Use File Explorer or PowerShell if migrating to a NAS or SAN. Use [AWS CLI](https://aws.amazon.com/cli/) if migrating to S3.

1. Detach `D:\`.

    !!!note
        At this point in the workflow your original SystemLink Server is in an operable state. Be aware if any new data is created or ingested by SystemLink Server at this time it will not be available to your new server.

1. Provision a new file store for SystemLink.

1. Install and configure the same or a newer version of SystemLink Server.

1. Configure SystemLink to use the newly created file storage location.

1. Install `nislmigrate` on your new SystemLink Server.

1. Attach the `D:\` volume used to capture data from your original SystemLink server.

1. Run the command `nislmigrate restore --all --secret <your secret> --change-file-store-root <new root> --dir D:\migration`.

    !!!note

>         The `--change-file-store-root` argument flag is needed to update the file meta data the new root location of your file storage. This could be a new drive letter, UNC path, or S3 URL depending on your configuration. If this step is not complete you will be able to view file meta data (name, properties, etc), but you will not be able to preview or download files.

Reviewer 3: Should we include examples here?

Author: I added some examples. @Christian-Nunnally can you check this and ensure this is what you expect?

>         The `--change-file-store-root` argument flag is needed to update the file meta data the new root location of your file storage. This could be a new drive letter, UNC path, or S3 URL depending on your configuration. If this step is not complete you will be able to view file meta data (name, properties, etc), but you will not be able to preview or download files.

Reviewer 2: You must use the `--change-file-store-root` argument flag to update the file meta data the new root location of your file storage. Otherwise, you will not be able to preview or download files. Depending on your configuration, this could be a new drive letter, UNC path, or S3 URL.

1. Verify your new SystemLink Server has all the expected migrated data.

### Single Node to Multi Node with PostGreSQL

This workflow describes steps to upgrade a single node deployment of SystemLink Server to a multi node deployment where the PostGreSQL instance used by the SystemLink Test Monitor service is hosted on a its own server or replica set.

> As of SystemLink 21.5.TODO VERSION NUMBER SystemLink supports using PostGreSQL as the database backing the Test Monitor service. This configuration is optional, but provides significant performance improvements. `nislmigrate` does not yet support migrating PostGreSQL directly.

Reviewer 3: Is this TODO still necessary?

Reviewer 3: I thought we just used PostgresSQL in 21.5, with no option for MongoDB for test data?

Author: Sorry, I hadn't pushed an upgrade after I got this version number from Nick. The todo should be gone now.

> The Test Monitor Service itself performs the migration of test steps, test results, and product from MongoDB to PostgreSQL. If you intend on utilizing a single node configuration the steps for this migration are the same as described in [**Single Node Migration**](#single-node-migration). Use the following workflow if you intend on migrating or upgrading to a new server and utilize a multi node configuration where PostgreSQL is hosted on a dedicated server or replica set.

Reviewer 2: The Test Monitor Service performs the migration of test steps, test results, and product from MongoDB to PostgreSQL. If you are using a single node configuration, follow the steps in [**Single Node Migration**](#single-node-migration). If  you intend on migrating or upgrading to a new server and utilize a multi node configuration where PostgreSQL is hosted on a dedicated server or replica set, use the following workflow.

> The Test Monitor Service itself performs the migration of test steps, test results, and product from MongoDB to PostGreSQL. If you intend on utilizing a single node configuration the steps for this migration are the same as described in [**Single Node Migration**](#single-node-migration). Use the following workflow if you intend on migrating or upgrading to a new server and utilize a multi node configuration where PostGreSQL is hosted on a dedicated server or replica set.

Reviewer 3: I'm not sure it's obvious in the text here that we only support migration from old-version (MongoDB) to new-version (PostgresSQL) here. This may need a Note callout.

Author: I adjusted the wording here a little, but I'm unsure if I've fully addressed your concern.

1. Backup your SystemLink Server.

1. Attach a volume to store data captured by `nislmigrate`. This will be referenced as `D:\`.

1. Install `nislmigrate`.

1. Run the command `nislmigrate capture --all --secret <your secret> --dir D:\migration`.

1. Detach `D:\`.

    !!!note
        At this point in the workflow your original SystemLink Server is in an operable state. Be aware if any new data is created or ingested by SystemLink Server at this time it will not be available to your new server.

1. Provision a new Windows server for SystemLink.

> 1. Provision a PostGreSQL server or replica set.

Reviewer 3: Unrelated, but I assume someone has the action item to write a chapter on Postgres for this handbook?

Author: I added a requirement to create this in the relevant feature regarding upgrading `nislmigrate` to support Postgres migration.

1. Install `nislmigrate` on your new SystemLink Server.

1. Attach the `D:\` volume used to capture data from your original SystemLink server.

1. Run the command `nislmigrate restore --all --secret <your secret> --dir D:\migration`

    !!!note
        After this step SystemLink will migrate your test steps, results, and products from MongoDB to PostGreSQL. Depending on the size of your data set this process may take some time.

1. Verify your new SystemLink Server has all the expected migrated data.

## Upgrading Multi Node Configurations

The following workflow describes upgrading a SystemLink Server instance that has been configured to use a dedicated MongoDB server or replica set, a dedicated PostGreSQL server or replica set, and a dedicated file store such as AWS S3.

1. Backup your SystemLink Server, MongoDB server or replica set, PostGreSQL server or replica set, and file store.

    !!!note
        If S3 is used as your file store the backup step is not needed since this managed internally by AWS.

1. Attach a volume to store data captured by `nislmigrate`. This will be referenced as `D:\`.

1. Install `nislmigrate`.

> 1. Run the command `nislmigrate capture --tags --dir D:\migration`.

Reviewer 3: What about --systems? We migrate SALT configuration.

Reviewer 3: --tags retrieves both current and historical data. Is it okay to replace the historical data? I guess because we drop the database things should be fine, but maybe another warning about not writing to the server during the migration?

Should we have a work item to add an option to only capture tag current values?

Reviewer 3: For very old version of the repository and system state services, their is file data stored on disk that is required for migration. This data will get copied into MongoDB for 21.x versions. We might need to include those services if we support migration from those versions.

Author: I added to the note regarding replacing tag history data. I do think we should add an option argument for tags that only migrates current value - FYI @Christian-Nunnally.

Per the old versions of states and repo service, I believe these are outside our support window for the migration tool at this time. I think we can add this, but I'll wait till we have customers requesting this first.

    !!!note
        While most of the data is external to SystemLink, tag current values are stored on the app server. Therefore this `nislmigrate` argument flag must be used to ensure that data is not lost during the migration.

1. Detach `D:\`.

    !!!note
        At this point in the workflow your original SystemLink Server is in an operable state. Be aware if any new data is created or ingested by the SystemLink server at this time it will not be available to your new server.

1. Provision a new Windows server for SystemLink.

1. Provision a MongoDB server or replica set from the backup previously created.

> 1. Provision a PostGres server or replica set from the backup previously created.

Reviewer: Not sure if the correct capitalization is `PostGres` or `Postgres`. Might be worthwhile to just use the full `PostgreSQL` here to avoid  confusion.

Author: Good catch! I thought I had found all instances where I was inconsistent.

    !!!note

>         The previous steps prescribe using a newly created instance of your MongoDB or PostgreSQL servers. This is because SystemLink may change the internal schema of these databases when services start post upgrade. If you do not start with a instance from a backup you will be unable to revert the MongoDB collections and PostgreSQL tables into the schemas needed for the older version of SystemLink.

Reviewer: ```suggestion
        The previous steps prescribe using a newly created instance of your MongoDB or PostgreSQL servers. This is because SystemLink may change the internal schema of these databases when services start post upgrade. If you do not start with an instance from a backup you will be unable to revert the MongoDB collections and PostgreSQL tables into the schemas needed for the older version of SystemLink.
```

1. Install and configure the new version of SystemLink Server.

1. Configure SystemLink to use the newly created MongoDB server or replica set, PostGreSQL server or replica set, and existing file store.

1. Install `nislmigrate` on your new SystemLink Server.

1. Attach the `D:\` volume used to capture data from your original SystemLink server.

1. Run the command `nislmigrate restore --tags --dir D:\migration`.

> 1. Verify your new SystemLink Server has all the expected migrated data.

Reviewer 3: It's a bit weird that we have separate sections for migrating to remote Mongo, remote files, and remote Postgres, but only a single section on migrating multi-node to multi-node? Should we encourage users to move all the data storage off the server for deployment? Do we need to talk about multi-node migrations where only part of the data is remote?

Specifically, we did work in the files service to allow you to migrate the files database with files stored on S3 or some other external location. We can migrate only the metadata in MongoDB without migrating files. Users will probably figure this out since trying to migrate files with files stored in S3 will just fail.

Author: You're correct that we want people to store all data off the app server. I'll add a note regarding that expectation, which is why I don't have a distinct section for each individual data store.

## Seamless Cut-over

> Since managed test systems are connected SystemLink it is desirable to approach upgrades or migrations such that test systems can connect to the new instance of SystemLink without manual intervention. To accomplish this the following conditions must be met.

Reviewer 2: Since managed test systems are connected SystemLink  test systems can connect to the new instance of SystemLink without manual intervention. To accomplish this the following conditions must be met.

To ensure that managed test systems can connect to the new instance of SystemLink without manual intervention (automatically?), check that the following conditions are met.

- Migration of systems data.

>     - This is either accomplished by using the `nislmigrate` argument flags `--all` or `--systems` or by connecting a MongoDB server or replica set that contains systems data.

Reviewer 3: I don't think the MongoDB setup will work. We need the captured SALT data to do a seemless transfer. If you just use the Mongo data, you will need to touch each system to clear out old key data and then re-approve.

Author: Good catch!

- The DNS name of the original SystemLink Server and the new SystemLink Server are the same.

> - If NI Web Server has been configured for HTTPS, the new SystemLink Server must use the same TLs certificate as the original SystemLink Server.

Reviewer 3: Instructions for migrating this certificate?

Reviewer 4: Do we need a certificate migrator?

Author: Let's leave this out of scope for now. Matt V mentioned that Josh has an example of how to do this in https://github.com/joshuaprewitt/systemlink-state-examples. I'm not sure which example does this. Regardless there is likely some cleanup work we would want to do on the example before providing this documentation. 

I created a new feature to cover the case of migrating a cert, or updating a cert and mass updating minions via a salt state. Lucky for us the need to support migration to K8s is going to encourage us to keep investing in the migration tool long term.

        | Migration destination | `nislmigrate` usage |
        | --------------------- | ------------------- |
        | New drive and directory | `nislmigrate restore --all --secret <your secret> --change-file-store-root X:\systemlink\data --dir D:\migration` |
        | UNC path | `nislmigrate restore --all --secret <your secret> --change-file-store-root \\FileShare\systemlink\data --dir D:\migration` |
        | AWS S3 | `nislmigrate restore --all --secret <your secret> --change-file-store-root s3://yours3bucket/systemlink/data/ --dir D:\migration` |

1. Verify your new SystemLink Server has all the expected migrated data.

### Single Node to Multi Node with PostgreSQL

Complete the following steps to upgrade a single node deployment of SystemLink Server to a multi node deployment where the PostgreSQL instance used by the SystemLink Test Monitor service is hosted on a its own server or replica set.

> As of SystemLink 21.3.2 SystemLink supports using PostgreSQL as the database backing the Test Monitor service. This configuration is optional, but provides significant performance improvements. `nislmigrate` does not yet support migrating between PostgreSQL servers or replica sets.

Reviewer 5: supports using PostgreSQL as an external database unless you install the side-band package for PostgreSQL that we would have to provide. The next full release will support a local installation of PostgreSQL from the main installer.

Author: I included a mention of future support for a single node configuration of Postgres

1. Provision a new Windows server for SystemLink.

1. Provision a PostgreSQL server or replica set.

1. Install and configure SystemLink 21.3.2.

1. Install `nislmigrate` on your new SystemLink Server.

1. Attach the `D:\` volume used to capture data from your original SystemLink server.

1. Run the command `nislmigrate restore --all --secret <your secret> --dir D:\migration`.

1. Open `C:\ProgramData\National Instruments\Skyline\Config\TestMonitor.json`.

1. Change the key `"Postgres.Enabled"` value from `false` to `true`.

1. Add the key `"Postgres.ConnectionString"` and a connection string value. For example:

        "Postgres.ConnectionString" : "Host=<insert db host>;Port=<insert port>;Username=<insert username>;Password=<insert password>;Database=<insert database>;SSL Mode=<Disable or Enable>"

1. Add the key `"Postgres.ConfigurationApproved"=true`.

1. Save and close the file.

1. Open the **NI SystemLink Server Configuration** application.

1. Navigate to **NI SystemLink Service Manager** and click **Restart**.

    !!!note

>         After this step SystemLink will migrate your test steps, results, and products from MongoDB to PostgreSQL. Depending on the size of your data set this process may take some time. The TestMonitor service will display a migration in progress status. If you see an error double check your connection string and restart services if you make any additional changes to the json file.

Reviewer 5: progress isn't available in the ServiceManager but is logged to log.txt on the server. Of note I've found that once you're in the section of the migration handling the TestSteps collection any run-rate estimates you perform are pretty accurate. Steps dominate the migration process and the migration is linear in execution (not exponential or logarithmic).

Author: I've included these relevant details in the !!!note.
